core/reactor_common.c

```diff
@@ -704,18 +749,23 @@ void _lf_pop_events() {
         // for which we decrement the reference count.
         if (event->trigger->token != event->token
                 && event->trigger->token != NULL) {
-            // Mark the previous one ok_to_free so we don't get a memory leak.
-            event->trigger->token->ok_to_free = OK_TO_FREE;
+            // Mark the previous one so we don't get a memory leak.
+            event->trigger->token->ok_to_free = token_and_value;
+            // Remove from the hashset to be freed at the end.
+            LF_PRINT_DEBUG("Removing %p from _lf_tokens_allocated", event->trigger->token);
+            hashset_remove(_lf_tokens_allocated, event->trigger->token);
```

Comment:
We should remove from _lf_tokens_allocated only when _lf_done_using method returns TOKEN_FREED.
Moreover
1. At all the places where we call hashset_remove(_lf_tokens_allocated, token); we should check on return value of _lf_done_using method.
2. There are some places in the code where we call _lf_done_using but is not paired with the call to remove from _lf_tokens_allocated hashset. As a defensive check (and to avoid double delete), we should call hashset_remove(_lf_tokens_allocated, token) when _lf_done_using returns TOKEN_FREED

Reply:
Thanks for the suggestion. However, _lf_tokens_allocated is rather specialized. It is used only at program termination and only so that any remaining tokens do not result in spurious memory leak reports from valgrind or similar tools.